### PR TITLE
TFP: 4L/192d at lr=4e-5 T_max=15 (depth + LR bracket)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
Vash (#2949) found 4L/192d/3H at lr=5e-5 reached test=0.002396, only +0.5% above baseline — still improving at timeout. This tests lr=4e-5 with T_max=15 (between T_max=10 baseline and T_max=20). Lower LR with intermediate schedule may find a deeper basin for the 4-layer architecture.

## Instructions
```bash
cd target/icml2026 && SENPAI_TIMEOUT_MINUTES=600 python train.py \
  --dataset tandemfoil_paper --optimizer lion --lr 4e-5 --cosine-t-max 15 \
  --grad-clip 0.5 --weight-decay 1e-2 --model-layers 4 --model-hidden-dim 192 \
  --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition --ema-decay 0.999 --epochs 999 \
  --eval-interval 10 --save-checkpoint --seed 0 \
  --wandb-group tfp-4l-lr-sweep
```
**Target:** val < 0.002344

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT yuji:

## Results

### Best Checkpoint (epoch 710)
| Metric | Baseline (3L, PR #3056) | 4L lr=4e-5 T_max=15 | Delta |
|--------|------------------------|---------------------|-------|
| **val/field_mse** | 0.002344 | **0.001903** | **-18.8%** |
| **test/field_mse** | 0.002221 | **0.001712** | **-22.9%** |
| val/surface_mse | — | 0.001108 | — |
| val/volume_mse | — | 0.001914 | — |
| test/surface_mse | — | 0.001018 | — |
| test/volume_mse | — | 0.001723 | — |

### Final EMA (epoch 778, at timeout)
| Metric | Value |
|--------|-------|
| val/field_mse | 0.001960 |
| test/field_mse | 0.001705 |

### Training Details
- **Epochs trained:** 778 (600-minute timeout)
- **Best epoch:** 710 (best val)
- **Peak VRAM:** ~37.7 GB (nvidia-smi during training)
- **W&B run:** [`gux6cdib`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/gux6cdib)
- **Seed:** 0

### Command
```bash
cd target/icml2026 && SENPAI_TIMEOUT_MINUTES=600 python train.py \
  --dataset tandemfoil_paper --optimizer lion --lr 4e-5 --cosine-t-max 15 \
  --grad-clip 0.5 --weight-decay 1e-2 --model-layers 4 --model-hidden-dim 192 \
  --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition --ema-decay 0.999 --epochs 999 \
  --eval-interval 10 --save-checkpoint --seed 0 \
  --wandb-group tfp-4l-lr-sweep --wandb-name "yuji/tfp-4l-lr4e5-tmax15" \
  --agent yuji
```

### What happened

**This is a clear and massive win.** The 4-layer/192-dim architecture at lr=4e-5 with T_max=15 cosine annealing dramatically outperforms the 3-layer baseline on both val and test.

Key observations:
1. **Slow start, powerful finish:** The 4L model lagged the 3L baseline until ~epoch 420, then crossed over and never looked back. The lower learning rate (4e-5 vs 1e-4) needed more epochs to converge but ultimately found a much better minimum.
2. **T_max=15 cosine cycles ratchet the EMA:** With EMA decay=0.999, each cosine trough (every 15 epochs) created opportunities to push the EMA model lower. Best val improved from 0.002035 (ep590) → 0.001963 (ep630) → 0.001941 (ep690) → 0.001903 (ep710).
3. **Extra depth pays off:** The 4th transformer layer provides additional representational capacity that the lower LR allows the optimizer to exploit gradually.
4. **Test generalizes better than val:** test/field_mse=0.001712 is -22.9% vs baseline test=0.002221, even better than the -18.8% val improvement — suggesting the 4L model is learning more robust representations.
5. **No code changes required:** This was a pure hyperparameter/architecture depth test; no modifications to train.py were needed.

### Suggested follow-ups
- **Try lr=3e-5 with 4L:** Even lower LR could push further, given that 4e-5 already dramatically improved on 1e-4.
- **Try 5L/192d at lr=4e-5:** If 4L is this much better than 3L, 5L may extract even more.
- **Try T_max=20:** Longer cosine cycles with the 4L model might allow even deeper troughs.
- **Run more seeds:** Verify this isn't a lucky seed — though the -22.9% test improvement is large enough to be confident.

**Comment by tcapelle** [comment]
ADVISOR: EXCELLENT finding! Your 4L/192d at lr=4e-5 T_max=15 produced **test/field_mse=0.001712 — NEW TFP TEST BEST** (beats haku baseline test=0.001789 by -4.3%). Val=0.001903 is slightly above val baseline (0.001857), which is why we can't merge yet.

**KEY INSIGHT:** 4L architecture at lower LR finds a better test basin. The 3L champion uses lr=7.5e-5 — the 4L model needs a different (lower) LR. We need to find the 4L LR sweet spot where BOTH val and test improve.

**NEXT TRIAL: lr=5e-5 at 4L T_max=15**

```bash
cd target/icml2026 && SENPAI_TIMEOUT_MINUTES=600 python train.py \
  --dataset tandemfoil_paper --optimizer lion --lr 5e-5 --cosine-t-max 15 \
  --grad-clip 0.5 --weight-decay 1e-2 --model-layers 4 --model-hidden-dim 192 \
  --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition --ema-decay 0.999 --epochs 999 \
  --eval-interval 10 --save-checkpoint --seed 0 \
  --wandb-group tfp-4l-lr-sweep --wandb-name "yuji/tfp-4l-lr5e5-tmax15" \
  --agent yuji
```

This probes lr=5e-5 — midpoint between your 4e-5 (test winner) and the 3L optimum at 7.5e-5. If val drops below 0.001857 we merge immediately.

**Targets:** val < 0.001857, test < 0.001712 (your current best)
